src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSourceProviders/DiagnosticSourceManager.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics.DiagnosticSources;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+[Export(typeof(IDiagnosticSourceManager)), Shared]
+internal sealed class DiagnosticSourceManager : IDiagnosticSourceManager
+{
+    /// <summary>
+    /// Document level <see cref="IDiagnosticSourceProvider"/> providers ordered by name.
+    /// </summary>
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _nameToDocumentProviderMap;
+
+    /// <summary>
+    /// Workspace level <see cref="IDiagnosticSourceProvider"/> providers ordered by name.
+    /// </summary>
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _nameToWorkspaceProviderMap;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public DiagnosticSourceManager([ImportMany] IEnumerable<IDiagnosticSourceProvider> sourceProviders)
+    {
+        _nameToDocumentProviderMap = sourceProviders
+            .Where(p => p.IsDocument)
+            .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+
+        _nameToWorkspaceProviderMap = sourceProviders
+            .Where(p => !p.IsDocument)
+            .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+    }
+
+    /// <inheritdoc />
```

Comment:
not needed.

Reply:
Please provide working code sample. Removing ToImmutableDictionary causes "cannot implicitly convert IEnumerable" error

Reply:
I think its the inheritdoc that isn't needed?